### PR TITLE
No setup in describe eslint

### PR DESCRIPTION
**Pull request by dhuang612**
completed refactoring for no-setup-in-describe

### CODE REVIEW

**Comment by dhuang612** [comment]
Hi,

yeah I'll fix it.

**Comment by dhuang612** [comment]
Also,

wanted to mention, for no setup in describe
```
 describe.skip()
```
is considered setup, but since these tests are going to be fixed by another commit.
 I simply did this:
```
// eslint-disable-next-line no-alert
```

File: core/test/regression/importer/importer_spec.js
Comment (gargol) [inline core/test/regression/importer/importer_spec.js]: this `should.exist(importer);` statement doesn't make sense :sweat_smile: something to be cleaned up after merge I think.

File: core/test/regression/site/dynamic_routing_spec.js
Comment (gargol) [inline core/test/regression/site/dynamic_routing_spec.js]: To clean up after merged ;)

File: core/test/unit/api/canary/utils/validators/input/pages_spec.js
Comment (gargol) [inline core/test/unit/api/canary/utils/validators/input/pages_spec.js]: Spaces need cleanup.

File: core/test/unit/api/canary/utils/validators/input/pages_spec.js
Comment (gargol) [inline core/test/unit/api/canary/utils/validators/input/pages_spec.js]: This function would always return `should fail for bad undefined`. The idea was to have a checked key returned for each iteration using `Object.keys(fieldMap).forEach((key) => {` :wink:

File: core/test/unit/api/canary/utils/validators/input/pages_spec.js
Comment (dhuang612) [reply core/test/unit/api/canary/utils/validators/input/pages_spec.js]: removed extra space and re-pushed PR

File: core/test/unit/api/canary/utils/validators/input/posts_spec.js
Comment (gargol) [inline core/test/unit/api/canary/utils/validators/input/posts_spec.js]: Same as above https://github.com/TryGhost/Ghost/pull/11154/files#r358046650 :wink:

File: core/test/unit/api/v2/utils/validators/input/pages_spec.js
Comment (gargol) [inline core/test/unit/api/v2/utils/validators/input/pages_spec.js]: Same here https://github.com/TryGhost/Ghost/pull/11154/files#r358046650. 
Thinking we might just create separate `it` statement per `key`. Could be something like:
```js
it(`should fail for bad slug`, function () {
    badValues = fieldMap.slug;
    checks = badValues.map((value) => {
        page = {};
        page[key] = value;

        frame = {
            options: {},
            data: {
                pages: [page]
            }
        };
        return validators.input.pages.add(apiConfig, frame)
            .then(Promise.reject)
            .catch((err) => {
                (err instanceof common.errors.ValidationError).should.be.true();
            });
    });
});
```

Or if you could think of a better approach? :)

File: core/test/unit/web/middleware/image/normalize_spec.js
Comment (gargol) [inline core/test/unit/web/middleware/image/normalize_spec.js]: The same problem with `undefined` here as in other cases ;)

File: core/test/unit/web/middleware/image/normalize_spec.js
Comment (gargol) [inline core/test/unit/web/middleware/image/normalize_spec.js]: New line :wink:

**Comment by gargol** [comment]
@dhuang612 the change looks good for the most part. Just need to figure out a better way to substitute the test cases which use iterations :) Lmk if you have any ideas about that or need help!

File: core/test/unit/api/v2/utils/validators/input/pages_spec.js
Comment (dhuang612) [reply core/test/unit/api/v2/utils/validators/input/pages_spec.js]: I'll stick to you're solution, thank you!

File: core/test/unit/web/middleware/image/normalize_spec.js
Comment (dhuang612) [reply core/test/unit/web/middleware/image/normalize_spec.js]: rewrote this, will resubmit for you to review

**Comment by dhuang612** [comment]
@gargol 

Hi,

I've made the changes you requested, please let me know if anything else needs to be updated.

**Comment by gargol** [comment]
Hey @dhuang612 I'll have to give it a little spin, but don't think I will be able to do it this year (the team is going on vacation starting this Friday). I'm keeping an eye on these changes :wink: Thanks for the contribution and hope you have a great holiday season!

**Comment by dhuang612** [comment]
Hi @gargol 
sorry what does "I'll have to give it a little spin mean"?
np, thanks for the update. 
Hope you have a great holiday season too!

**Comment by gargol** [comment]
> do you mean do some testing

Yes, exactly ;) Just need to verify if the tests are failing we still keep cases readable.

File: core/test/unit/web/middleware/uncapitalise_spec.js
Comment (gargol) [inline core/test/unit/web/middleware/uncapitalise_spec.js]: @dhuang612 don't think this way of iteration over the `apiVersions` would work in this case (and all the other cases similar to this). The reason being, `done()` will be called during the first iteration of the `it` block and only `v0.1` would be checked. 

After thinking about this problem for a bit I think it's most reasonable just to add an es-lint ignore rule for `no-setup-in-describe` rule with a short note explaining the reason for ignore. Something along the lines:
```
// eslint-disable no-useless-escape
// NOTE: there is no good way to substitute dynamic test setup based on an array of values, see {link to this PR}
```

**Review by naz — CHANGES_REQUESTED** [review]

File: core/test/unit/web/middleware/uncapitalise_spec.js
Comment (gargol) [inline core/test/unit/web/middleware/uncapitalise_spec.js]: @dhuang612 awesome :smile: I think, this is exactly how we should deal with these types of test case iterations :+1: 

Think all is left in this PR is to apply (or rather revert) changes to other places where we iterate over multiple test cases and leave a similar comment there. Think we are very close to merging this gigantic change!

File: core/test/unit/web/middleware/uncapitalise_spec.js
Comment (gargol) [inline core/test/unit/web/middleware/uncapitalise_spec.js]: I've noticed you have these stray empty space symbols left in few places. One great config change you could make in the editor you are using is enabling "Trim Trailing Whitespace" (at least that's how it's called in VS Code) :wink:

File: core/test/unit/web/middleware/uncapitalise_spec.js
Comment (dhuang612) [reply core/test/unit/web/middleware/uncapitalise_spec.js]: Hey @gargol sorry I've been really sick and resting. Just starting to get back into finishing up this pr. 

I was wondering, when you talk about the other test case iterations that need to be reverted back and commented. Are you talking about these?
**pages_spec.js**
```
describe('field formats', function () {
            const fieldMap = {
                title: [123, new Date(), _.repeat('a', 2001)],
                slug: [123, new Date(), _.repeat('a', 192)],
                mobiledoc: [123, new Date()],
                feature_image: [123, new Date(), 'random words'],
                featured: [123, new Date(), 'abc'],
                status: [123, new Date(), 'abc'],

```

File: core/test/unit/web/middleware/uncapitalise_spec.js
Comment (dhuang612) [reply core/test/unit/web/middleware/uncapitalise_spec.js]: I have enabled this setting under visual studio code but it's not working.
Researching into next steps.

Comment (dhuang612) [reply core/test/unit/web/middleware/uncapitalise_spec.js]: so after reviewing this more, under .editorconfig file
```
[*]
insert_final_newline = true
```

is adding in the extra new line is it not?

**Comment by daniellockyer** [comment]
Hey @dhuang612! Would love to get this merged at some point. If you sort out the whitespace issue and rebase on top of `master`, then we can see where we're at :smile: Thanks!

**Comment by dhuang612** [comment]
@daniellockyer 

Hey,

I've been looking into getting these last items addressed.

Running into an issue here, I've rebased my local master to you're guys branch.
on GitHub my master project is showing even with you guys

Screen Shot 2020-03-09 at 2 41 04 PM


However, under git status I'm showing two files having been changed that I can't seem to update?

Screen Shot 2020-03-09 at 2 38 32 PM


Screen Shot 2020-03-09 at 2 38 32 PM

With these two files not updating, when I go to  no-setup-in-describe-eslint branch and update to my local master it then complains about not being able to merge the commits that were changes to the files mentioned above.

I haven't seen this before, hoping to get some help.

Thanks, excited to finish up this PR!

**Comment by gargol** [comment]
Hey @dhuang612! You can ignore those two files being changed and use `git reset core/client` and `git reset content/themes/casper` on you local environment :wink: It's just an issue of linked repositories going out of sync after Ghost master goes past a version bump.

**Comment by dhuang612** [comment]
@gargol Hey
I finally had a chance to look at this again. 
Thank you for you're help in resolving those conflicts.

After the reset followed instructions from StackOverflow to remove them completely.
Took a look at the white space issue and that should also be solved.
Please let me know if there is anything else to resolve.

File: .gitmodules
Comment (ErisDS) [inline .gitmodules]: This must not be removed

Comment (dhuang612) [reply .gitmodules]: @ErisDS  resolved.

**Comment by allouis** [comment]
Closing for inactivity, feel free to reopen once you've rebased and finished this work!

**Comment by dhuang612** [comment]
excuse me, I was waiting for a response from @gargol

**Pull request closed without merging**
